src/include/duckdb/execution/index/art/art_node.hpp

```diff
@@ -0,0 +1,146 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/execution/index/art/art_node.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/execution/index/art/fixed_size_allocator.hpp"
+#include "duckdb/execution/index/art/swizzleable_pointer.hpp"
+
+namespace duckdb {
+
+// classes
+enum class ARTNodeType : uint8_t {
+	PREFIX_SEGMENT = 1,
+	LEAF_SEGMENT = 2,
+	LEAF = 3,
+	NODE_4 = 4,
+	NODE_16 = 5,
+	NODE_48 = 6,
+	NODE_256 = 7
+};
+class ART;
+class ARTNode;
+class Prefix;
+class MetaBlockReader;
+class MetaBlockWriter;
+
+// structs
+struct BlockPointer;
+
+//! The ARTNode is the swizzleable pointer class of the ART index.
+//! If the ARTNode pointer is not swizzled, then the leftmost byte identifies the ARTNodeType.
+//! The remaining bytes are the position in the respective ART buffer.
+class ARTNode : public SwizzleablePointer {
+public:
+	// constants (this allows testing performance with different ART node sizes)
+	// node prefixes (NOTE: this should always hold: PREFIX_SEGMENT_SIZE >= PREFIX_INLINE_BYTES)
+	static constexpr uint32_t PREFIX_INLINE_BYTES = 8;
+	static constexpr uint32_t PREFIX_SEGMENT_SIZE = 32;
+	// node thresholds
+	static constexpr uint8_t NODE_48_SHRINK_THRESHOLD = 12;
+	static constexpr uint8_t NODE_256_SHRINK_THRESHOLD = 36;
+	// node sizes
+	static constexpr uint8_t NODE_4_CAPACITY = 4;
+	static constexpr uint8_t NODE_16_CAPACITY = 16;
+	static constexpr uint8_t NODE_48_CAPACITY = 48;
+	static constexpr uint16_t NODE_256_CAPACITY = 256;
+	// others
+	static constexpr uint8_t EMPTY_MARKER = 48;
+	static constexpr uint32_t LEAF_SEGMENT_SIZE = 8;
+
+public:
+	//! Constructs an empty ARTNode
+	ARTNode();
+	//! Constructs a swizzled pointer from a block ID and an offset
+	explicit ARTNode(MetaBlockReader &reader);
+	//! Get a new pointer to a node, might cause a new buffer allocation, and initialize it
+	static void New(ART &art, ARTNode &node, const ARTNodeType &type);
```

Comment:
Passing something as a reference passes it as a pointer. Since `ARTNodeType` is a 1-byte field, passing an 8-byte pointer to a 1-byte field is worse from a performance perspective than just passing an `ARTNodeType` directly. As a general rule you should only use pass-by-reference when passing fields that are larger than 8 bytes or that involve non-trivial copies.